packages/core/lib/patchers/aws3_p.ts

```diff
@@ -62,110 +82,113 @@ function addFlags(http: HttpResponse, subsegment: Subsegment, err?: SdkError): v
-  // create local copy so that we can later call it
-  const send = client.send;
-
-  const serviceIdentifier = client.config.serviceId;
-
-  client.send = async (...args: Parameters<typeof client.send>): Promise<Output> => {
-    const [command] = args;
-    const segment = manualSegment || contextUtils.resolveSegment();
-
-    let operation: string = command.constructor.name.slice(0, -7);
-
-    if (!segment) {
-      const output = serviceIdentifier + '.' + operation;
-
-      if (!contextUtils.isAutomaticMode()) {
-        logger.getLogger().info('Call ' + output + ' requires a segment object' +
-          ' on the request params as "XRaySegment" for tracing in manual mode. Ignoring.');
-      } else {
-        logger.getLogger().info('Call ' + output +
-          ' is missing the sub/segment context for automatic mode. Ignoring.');
-      }
-      return send.apply(client, args) as Promise<Output>;
+const getXRayMiddleware = (config: RegionResolvedConfig, manualSegment?: SegmentLike): BuildMiddleware<any, any> => (next: any, context: any) => async (args: any) => {
+  const segment = contextUtils.isAutomaticMode() ? contextUtils.resolveSegment() : manualSegment;
+  const {clientName, commandName} = context;
+  const operation: string = commandName.slice(0, -7); // Strip trailing "command" string
+  const service: string = clientName.slice(0, -6);    // Strip trailing "client" string
+
+  if (!segment) {
+    const output = service + '.' + operation.charAt(0).toLowerCase() + operation.slice(1);
+
+    if (!contextUtils.isAutomaticMode()) {
+      logger.getLogger().info('Call ' + output + ' requires a segment object' +
+        ' on the request params as "XRaySegment" for tracing in manual mode. Ignoring.');
+    } else {
+      logger.getLogger().info('Call ' + output +
+        ' is missing the sub/segment context for automatic mode. Ignoring.');
     }
+    return next(args);
+  }
 
-    const subsegment = segment.addNewSubsegment(serviceIdentifier);
-    subsegment.addAttribute('namespace', 'aws');
+  const subsegment: Subsegment = segment.addNewSubsegment(service);
+  subsegment.addAttribute('namespace', 'aws');
 
-    try {
-      const res = (await send.apply(client, [command])) as Output;
+  const parent = (segment instanceof Subsegment
+    ? segment.segment
+    : segment);
 
-      if (!res) throw new Error('Failed to get response from instrumented AWS Client.');
+  args.request.headers['X-Amzn-Trace-Id'] = stringify(
+    {
+      Root: parent.trace_id,
+      Parent: subsegment.id,
+      Sampled: parent.notTraced ? '0' : '1',
+    },
+    ';',
+  );
 
+  let res;
+  try {
+    res = await next(args);
+    if (!res) throw new Error('Failed to get response from instrumented AWS Client.');
+
+    // TODO: Figure out how to get region string. Should look like 'await options.region()'
+    const [aws, http] = await buildAttributesFromMetadata(
+      service,
+      operation,
+      await config.region(),
```

Comment:
Is this an appropriate way to get region, from the client's `RegionResolvedConfig`?

Reply:
Yes. Here is an example of how it's done in AWS SDK for JavaScript (v3) https://github.com/aws/aws-sdk-js-v3/blob/2165d4a5c2aec2945900bbe8e79d497da8a80cfa/packages/config-resolver/src/EndpointsConfig.ts#L47-L49